power_grid_model_c/power_grid_model/include/power_grid_model/main_core/update.hpp

```diff
@@ -13,29 +13,48 @@
 namespace power_grid_model::main_core {
 
 namespace detail {
-template <std::derived_from<Base> Component, class ComponentContainer, std::forward_iterator ForwardIterator,
-          typename Func>
+template <std::derived_from<Base> Component, class ComponentContainer, typename UpdateType>
     requires model_component_state<MainModelState, ComponentContainer, Component> &&
-             std::invocable<std::remove_cvref_t<Func>, typename Component::UpdateType, Idx2D const&>
-inline void iterate_component_sequence(Func&& func, MainModelState<ComponentContainer> const& state,
-                                       ForwardIterator begin, ForwardIterator end,
+             std::same_as<std::remove_cvref_t<typename Component::UpdateType>, std::remove_cvref_t<UpdateType>>
+inline Idx2D get_idx_by_id(MainModelState<ComponentContainer> const& state, UpdateType const& update) {
+    return state.components.template get_idx_by_id<Component>(update.id);
+}
+
+template <std::derived_from<Base> Component, std::forward_iterator ForwardIterator, typename Func>
+    requires std::invocable<std::remove_cvref_t<Func>, typename Component::UpdateType, Idx2D const&>
+inline void iterate_component_sequence(Func&& func, ForwardIterator begin, ForwardIterator end,
                                        std::vector<Idx2D> const& sequence_idx) {
-    bool const has_sequence_id = !sequence_idx.empty();
     Idx seq = 0;
 
     // loop to to update component
     for (auto it = begin; it != end; ++it, ++seq) {
-        // get component
-        // either using ID via hash map
-        // either directly using sequence id
-        Idx2D const sequence_single =
-            has_sequence_id ? sequence_idx[seq] : state.components.template get_idx_by_id<Component>(it->id);
```

Comment:
for all update types, we now determine the sequence beforehand.

The rationale is:

* for independent batch updates, the sequence is determined once
  * this is the same behaviour as before
* for non-independent batch updates, the sequence is determined for every scenario separately
  * this behaviour is equivalent to the situation before, but with half the amount of ID lookups
* for permanent updates, the sequence is determined once
  * this only causes a minor memory overhead compared to the situation before, but that is not the bottleneck.
  * in addition, permanent updates are less performance-critical, so it's not an issue

this allows this piece of complexity to be removed.